### PR TITLE
NO-ISSUE: Abstract how to get spoke kubeconfig from ClusterReference and add tests

### PR DESCRIPTION
**Pull request by javipolo**
Abstract the way that we're getting the spoke Admin KubeConfig name
This way it can be reused if needed and we can also run some tests against it

/cc @carbonin

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [x] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] Unit tests
- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 1]
```diff
@@ -3519,7 +3519,42 @@ var _ = Describe("selectClusterNetworkType", func() {
 	}
 })
 
-var _ = Describe("Getting ClusterDeployment admin kubeconfig secret name", func() {
+var _ = Describe("Getting spoke admin kubeconfig secret name from ClusterReference", func() {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:3522]: Usually this should be the name of the function you're testing.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 2]
```diff
@@ -3519,7 +3519,42 @@ var _ = Describe("selectClusterNetworkType", func() {
...
+		c = fakeclient.NewClientBuilder().WithScheme(scheme.Scheme).Build()
+		cd = newClusterDeployment(clusterName, testNamespace, defaultClusterSpec)
+		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{
+			AdminKubeconfigSecretRef: corev1.LocalObjectReference{
+				Name: adminSecretName,
+			},
+		}
+	})
+	Context("ClusterDeployment does not exist", func() {
+		It("can make secret name from template", func() {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:3544]: There's no need for a context when it contains a single test and doesn't have any setup code.

File: internal/controller/controllers/clusterdeployments_controller_test.go [hunk 3]
```diff
@@ -3519,7 +3519,42 @@ var _ = Describe("selectClusterNetworkType", func() {
...
+			Expect(getClusterReferenceAdminKubeConfigSecretName(ctx, c, &clusterReference)).To(Equal(fmt.Sprintf(adminKubeConfigStringTemplate, cd.Name)))
+		})
+	})
+	Context("ClusterDeployment exists", func() {
+		It("can extract secret name from ClusterDeployment", func() {
```
Comment (carbonin) [inline internal/controller/controllers/clusterdeployments_controller_test.go:3549]: Same here, this should not have a context and be named something like:
`It("gets the secret name from the cluster deployment when it exists", func() {`

**Review by carbonin — COMMENTED** [review]
Just some comments on the test names, but I'm curious what the blocker is to testing the function as it existed before.

We should be able to use the test client object for the queries and a mock client factory, right?
I don't see a great reason to break this function apart outside of testing so if we can do it without breaking it up that would probably be better.

**Comment by javipolo** [comment]
> Just some comments on the test names, but I'm curious what the blocker is to testing the function as it existed before.
> 
> We should be able to use the test client object for the queries and a mock client factory, right? I don't see a great reason to break this function apart outside of testing so if we can do it without breaking it up that would probably be better.

My main reason to start the splitting was to not have logics on creating the default secretname in two different files of the code:

https://github.com/openshift/assisted-service/blob/268172d230603597fe28a6de4c329a072833b2bc/internal/controller/controllers/agent_controller.go#L284
https://github.com/openshift/assisted-service/blob/268172d230603597fe28a6de4c329a072833b2bc/internal/controller/controllers/clusterdeployments_controller.go#L2059

Once there, I thought that maybe it would be better to split the way I did, putting all logic at least in the same file, and it would also accomodate for nicer testing and possibly reusing that code if needed

Let me know what do you think, but if you feel that not splitting the functions is better I'll rewrite the tests to accomodate the situation :)

**Comment by carbonin** [comment]
>My main reason to start the splitting was to not have logics on creating the default secretname in two different files of the code

If this is the reason, then I'd expect the new function to be called from multiple places, but it seems it's only called from one. It looks like you moved the duplicated code so it's co-located, but didn't deduplicate.

File: internal/controller/controllers/agent_controller_test.go [hunk 4]
```diff
@@ -3135,3 +3136,112 @@ var _ = Describe("TestConditions", func() {
...
+		Expect(err).To(HaveOccurred())
+	})
+
+	It("uses the kubeconfig template format when the cluster deployment is missing", func() {
+		cdSpec := hivev1.ClusterDeploymentSpec{
+			ClusterMetadata: &hivev1.ClusterMetadata{},
+		}
+
+		cd := newClusterDeployment(clusterName, testNamespace, cdSpec)
+		Expect(c.Create(ctx, cd)).To(Succeed())
```
Comment (carbonin) [inline internal/controller/controllers/agent_controller_test.go:3225]: Shouldn't this not be created here (just based on the test name)?

Comment (javipolo) [reply internal/controller/controllers/agent_controller_test.go]: You're totally right :sweat_smile: fixed!

**Review by carbonin — APPROVED** [review]

**Pull request merged**